client/src/console.rs

```diff
@@ -0,0 +1,234 @@
+                    message.sender
+                )
+                .chars()
+                .enumerate()
+                .flat_map(|(i, chr)| {
+                    if i != 0 && i % screen.get_width() as usize == 0 {
+                        lines += 1;
+                        Some('\n')
+                    } else {
+                        None
+                    }
+                    .into_iter()
+                    .chain(std::iter::once(chr))
+                })
+                .collect::<String>();
+                screen.print_fbg(x, y, &text, THEME_FG, THEME_BG);
+                lines
+            }
+            Message::System(message) => {
+                let mut lines = 1;
+                let message = &message
+                    .chars()
+                    .enumerate()
+                    .flat_map(|(i, chr)| {
+                        if i != 0 && i % screen.get_width() as usize == 0 {
+                            lines += 1;
+                            Some('\n')
+                        } else {
+                            None
+                        }
+                        .into_iter()
+                        .chain(std::iter::once(chr))
+                    })
+                    .collect::<String>();
+                screen.print_fbg(x, y, message, Color::DarkGrey, THEME_BG);
+                lines
+            }
+            Message::Error(message) => {
+                let mut lines = 1;
+                let message = &message
+                    .chars()
+                    .enumerate()
+                    .flat_map(|(i, chr)| {
+                        if i != 0 && i % screen.get_width() as usize == 0 {
+                            lines += 1;
+                            Some('\n')
+                        } else {
+                            None
+                        }
+                        .into_iter()
+                        .chain(std::iter::once(chr))
+                    })
+                    .collect::<String>();
+                screen.print_fbg(x, y, message, Color::Red, THEME_BG);
+                lines
+            }
+        }
+    }
+}
+
+pub struct UserListWindow {
+    screen: Screen,
+    dirty: bool,
+    user_list: BTreeSet<String>,
+}
+
+impl UserListWindow {
+    pub fn new(w: u32, h: u32) -> Self {
+        Self {
+            screen: Screen::new(w, h),
+            user_list: BTreeSet::new(),
+            dirty: true,
+        }
+    }
+
+    pub fn resize(&mut self, w: u32, h: u32) {
+        self.dirty = true;
+        self.screen.resize(w, h)
+    }
+
+    pub fn set_list(&mut self, new_list: Vec<String>) {
+        self.user_list.clear();
+        for entry in new_list {
+            self.add_user(entry)
+        }
+    }
+
+    pub fn add_user(&mut self, username: String) {
+        if !self.user_list.contains(&username) {
+            self.dirty = true;
+            self.user_list.insert(username);
+        }
```

Comment:
How about:
```rust
self.dirty |= !self.user_list.insert(username);
```
?
This should behave the same, because `insert` already checks if `user_list` contains the value (and returns if it was there).